zhusuan/distributions/multivariate.py

```diff
@@ -952,3 +953,213 @@ def _prob(self, given):
+
+    def _get_value_shape(self):
+        shape_ = tf.TensorShape([
+            self._n_in if isinstance(self._n_in, int) else None,
+            self._n_out if isinstance(self._n_out, int) else None])
+        return shape_
+
+    def _batch_shape(self):
+        return tf.shape(self.mean)[:-2]
+
+    def _get_batch_shape(self):
+        if self.mean.get_shape():
+            return self.mean.get_shape()[:-2]
+        return tf.TensorShape(None)
+
+    def _sample(self, n_samples):
+        mean, u_tril, v_tril = self.mean, self.u_tril, self.v_tril
+        if not self.is_reparameterized:
+            mean = tf.stop_gradient(mean)
+            u_tril = tf.stop_gradient(u_tril)
+            v_tril = tf.stop_gradient(v_tril)
+
+        def tile(t):
+            new_shape = tf.concat([[n_samples], tf.ones_like(tf.shape(t))], 0)
+            return tf.tile(tf.expand_dims(t, 0), new_shape)
+
+        def transpose_last2_dims(t):
+            len_ = len(t.get_shape())
+            return tf.transpose(t, [i for i in range(len_-2)]+[len_-1, len_-2])
+
+        batch_mean = tile(mean)
+        batch_u_tril = tile(u_tril)
+        batch_v_tril = tile(v_tril)
+        noise = tf.random_normal(tf.shape(batch_mean), dtype=self.dtype)
+        samples = batch_mean + \
+            tf.matmul(tf.matmul(batch_u_tril, noise),
+                      transpose_last2_dims(batch_v_tril))
+        # Update static shape
+        static_n_samples = n_samples if isinstance(n_samples, int) else None
+        samples.set_shape(tf.TensorShape([static_n_samples])
+                          .concatenate(self.get_batch_shape())
+                          .concatenate(self.get_value_shape()))
+        return samples
+
+    def _log_prob(self, given):
+        mean, u_tril, v_tril = (self.path_param(self.mean),
+                                self.path_param(self.u_tril),
+                                self.path_param(self.v_tril))
+
+        log_det_u = 2 * tf.reduce_sum(
+            tf.log(tf.matrix_diag_part(u_tril)), axis=-1)
+        log_det_v = 2 * tf.reduce_sum(
+            tf.log(tf.matrix_diag_part(v_tril)), axis=-1)
+        n_in = tf.cast(self._n_in, self.dtype)
+        n_out = tf.cast(self._n_out, self.dtype)
+        logZ = - n_in * n_out / 2 * \
+            tf.log(2 * tf.constant(np.pi, dtype=self.dtype)) - \
+            n_in / 2 * log_det_v - n_out / 2 * log_det_u
+        # logZ.shape == batch_shape
+        if self._check_numerics:
+            logZ = tf.check_numerics(logZ, "log[det(Cov)]")
+
+        def transpose_last2_dims(t):
+            len_ = len(t.get_shape())
+            return tf.transpose(
+                t, [i for i in range(len_-2)]+[len_-1, len_-2])
+
+        y = given - mean
+        y_with_last_dim_changed = tf.expand_dims(tf.ones(tf.shape(y)[:-1]), -1)
+        Lu, _ = maybe_explicit_broadcast(
+            u_tril, y_with_last_dim_changed,
+            'MultivariateNormalCholesky.u_tril', 'expand_dims(given, -1)')
```

Comment:
why is this broadcast?

Reply:
y can have 1 additional sample dimension.

Reply:
I see. Though note that actually `given` can have >=1 additional sample dimensions.